turbopack/crates/turbopack-ecmascript/src/analyzer/side_effects.rs

```diff
@@ -0,0 +1,2517 @@
+                } else {
+                    // Unknown constructor calls are considered to have side effects
+                    self.mark_side_effect();
+                }
+            }
+            Expr::Assign(_) => {
+                // Assignments have side effects
+                // TODO: allow assignments to module level variables
+                self.mark_side_effect();
+            }
+            Expr::Update(_) => {
+                // Updates (++, --) have side effects
+                // TODO: allow updates to module level variables
+                self.mark_side_effect();
+            }
+            Expr::Await(_) | Expr::Yield(_) => {
+                // Async operations have side effects, this could be less conservative
+                self.mark_side_effect();
+            }
+            Expr::TaggedTpl(tagged_tpl) => {
+                // Tagged template literals are function calls
+                // But some are known to be pure, like String.raw
+                if self.is_known_pure_builtin_function(&tagged_tpl.tag) {
+                    for arg in &tagged_tpl.tpl.exprs {
+                        arg.visit_with(self);
+                    }
+                } else {
+                    self.mark_side_effect();
+                }
+            }
+            Expr::OptChain(opt_chain) => {
+                // Optional chaining can be pure if it's just member access
+                // But if it's an optional call, it has side effects
+                opt_chain.base.visit_with(self);
+            }
+            Expr::Seq(seq) => {
+                // Sequence expressions - check each expression
+                for expr in &seq.exprs {
+                    expr.visit_with(self);
+                }
+            }
+            Expr::SuperProp(super_prop) => {
+                // Super property access is pure (reading from parent class)
+                // Check if the property expression has side effects
+                super_prop.prop.visit_with(self);
+            }
+            Expr::MetaProp(_) => {
+                // Meta properties like import.meta and new.target are pure
+                // They just read metadata, don't cause side effects
+            }
+            Expr::JSXMember(_) | Expr::JSXNamespacedName(_) | Expr::JSXEmpty(_) => {
+                // JSX member expressions and names are pure (they're just identifiers)
+            }
+            Expr::JSXElement(_) | Expr::JSXFragment(_) => {
+                // JSX elements compile to function calls (React.createElement, etc.)
+                // These are side-effectful unless we know the JSX factory is pure
+                self.mark_side_effect();
+            }
+            Expr::PrivateName(_) => {
+                // Private names are pure (just identifiers)
+            }
+
+            // Be conservative for other expression types
+            // To support more
+            _ => {
+                expr.visit_children_with(self);
```

Comment:
```suggestion
                self.mark_side_effect();
```

For unknown expression it would be safer to mark them as side effect.